metriq_gym/helpers/qedc_helpers.py

```diff
@@ -0,0 +1,188 @@
+        counts_list: a list of all counts objects, each index corresponds to a circuit.
+
+    Returns:
+        circuit_metrics: the updated circuit metrics in QED-C's format.
+    """
+
+    class CountsWrapper:
+        """
+        A wrapper class to enable support with QED-C's method to analyze results.
+        """
+
+        def __init__(self, counts: dict[str, int]):
+            self.counts = counts
+
+        def get_counts(self, _):
+            return self.counts
+
+    # Import the correct module
+    benchmark_name = str(params["benchmark_name"])
+    benchmark = import_benchmark_module(benchmark_name)
+
+    # Initialize metrics module using qiskit in QED-C submodule.
+    benchmark.qedc_benchmarks_init(api="qiskit")
+
+    # Restore circuit metrics dictionary from the dispatch data
+    metrics.circuit_metrics = job_data.circuit_metrics
+
+    # Iterate and get the metrics for each circuit in the list.
+    for curr_idx, (num_qubits, s_str) in enumerate(job_data.circuit_identifiers):
+        counts: dict[str, int] = counts_list[curr_idx]
+
+        result_object = CountsWrapper(counts)
+
+        if benchmark_name == QEDC_Benchmark_Names.PHASE_ESTIMATION:
+            # Requires slightly different arguments.
+            _, fidelity = benchmark.analyze_and_print_result(
+                None, result_object, int(num_qubits) - 1, float(s_str), params["num_shots"]
+            )
+
+        elif benchmark_name == QEDC_Benchmark_Names.QUANTUM_FOURIER_TRANSFORM:
+            # Requires slightly different arguments.
+            _, fidelity = benchmark.analyze_and_print_result(
+                None,
+                result_object,
+                int(num_qubits),
+                int(s_str),
+                params["num_shots"],
+                params["method"],
+            )
+
+        else:
+            # Default call for Bernstein-Vazirani and Hidden Shift.
+            _, fidelity = benchmark.analyze_and_print_result(
+                None, result_object, int(num_qubits), int(s_str), params["num_shots"]
+            )
+
+        metrics.store_metric(num_qubits, s_str, "fidelity", fidelity)
+
+    return metrics.circuit_metrics
+
+
+def get_circuits_and_metrics(
+    params: dict[str, float | str],
+) -> tuple[list[QuantumCircuit], QEDC_Metrics, list[tuple[str, str]]]:
+    """
+    Uses QED-C submodule to obtain circuits and circuit metrics.
+
+    Args:
+        params: the parameters to run the benchmark with, also includes benchmark_name.
+
+    Returns:
+        circuits: the list of quantum circuits for the benchmark.
+        circuit_metrics: the circuit metrics at the time of circuit creation.
+        circuit_identifiers: the unique identifiers for each circuit (num qubits, secret str).
+    """
+
+    # Import the correct module
+    benchmark = import_benchmark_module(str(params["benchmark_name"]))
+
+    # Remove benchmark_name so it can be passed into the run call
+    params.pop("benchmark_name", None)
```

Comment:
Perhaps there is a cleaner version of doing this that avoids `pop()`. After all, `params` comes from a model dump of a pydantic object, and you can actually do `model_dump(exclude=...` in Pydantic.